### DISCUSSION TITLE
Remove border when window is the only one on screen

### DISCUSSION THREAD

**Discussion by Schweber** [YaLTeR/niri#508] [Ideas]
I'd like the focus-ring and border to go away when there is only one window visible on screen (no matter the matter the reason: maximized, the only window at all etc.).

**Comment by YaLTeR** [comment]
Hm, I guess there could be a window rule matcher about the column being maximized, but I'm not sure this behavior in general is necessarily good. You could instead try fullscreening?

**Reply by Schweber** [reply]
I could but then it would also cover the status bar (e.g. `waybar`). I find borders useful to differentiate between multiple windows and seeing which one has focus. But if only one window is on screen, the border is distracting and annoying to me.

`hyprland` has this option `no_gaps_when_only` which i'm looking to replicate in `niri`. This way, the status bar stays visible while the window takes up all the space and no border is drawn.

**Reply by Dimev** [reply]
hyprland's option here is nice.
I think it could also be implemented as a "gaps-when-maximized" option?

**Reply by zefr0x** [reply]
Hyprland removed this option in the latest version.

Window rules and workspace rules are flexible to the point that it cloud be done using them:
https://wiki.hyprland.org/0.45.0/Configuring/Workspace-Rules/#smart-gaps

**Reply by Schweber** [reply]
Sure but how you do it in niri right now?

**Reply by zefr0x** [reply]
> Sure but how you do it in niri right now?

i don't

**Comment by Dimev** [comment]
How hard is it to add extra window rules?
I might take a look at contributing one for a maximized column.

**Reply by YaLTeR** [reply]
It's not hard to add extra window rules, but it may be hard to implement the specific logic that some window rule needs. It's a bit unclear what exactly the design should be for this discussion. A window rule for a sole window on a workspace is easy. For maximized active window shouldn't be hard, though there's an edge case of what should happen if the window refuses to expand to the full size (i.e. if you try to maximize a fixed size dialog). Should the rule still apply or not? If yes then it's simple, if not then you can get into all sorts of edge cases.

**Reply by Dimev** [reply]
Yeah that would make sense, probably wouldn't want the borders to be removed if it doesn't maximize.

I'd also like it to be able to disable gaps. Essentially just a fullscreen without also hiding the bar.

**Comment by gmankab** [comment]
i am switching to niri from hyprland, and i really miss no_gaps_when_only analog in niri

**Reply by maikelthedev** [reply]
Me too. It's the one missing feature.

**Comment by benjajaja** [comment]
Same here, basically any of these options would work for me:

* Allow showing status bar (e.g. waybar) also in fullscreen mode.
* When a maximized windows is the only window showing (i.e. not offset / sharing screen with other columns), then hide focus-ring and border.
* When workspace has only one window, then hide focus-ring and border.
* Add a key binding to toggle borders on/off for the focused/active window.

Could be window rules, if there was a match for either of the first two options. However, I imagine that matching by a workspace property (e.g. "only one window", or even "workspace name") could be problematic, as you can immediately set those properties in the same rule.

Right now I'm using this as a crutch:
```
window-rule {                                                                                             
    match app-id="firefox"                                                                                

    open-maximized true                                                                                   

    border {                                                                                              
        off                                                                                               
    }                                                                                                     
}
```

But with the severe limitation that it never shows a border for this specific app. If I open another firefox window for example, it should then show borders (unless I also open it in an empty workspace). Especially e.g. terminals can't work with this.

**Reply by benjajaja** [reply]
I'm coming from this xmonad setting: https://xmonad.github.io/xmonad-docs/xmonad-contrib/XMonad-Layout-NoBorders.html#v:smartBorders

**Reply by maikelthedev** [reply]
seems to be pretty much the same as no_gaps_when_only in hyprland

**Comment by maikelthedev** [comment]
This doesn't seem to be a valid option though. 

```kdl
window-rule {                                                                                             
    match app-id="librewolf"

    open-maximized true                                                                                   

    gaps {                                                                                              
        off                                                                                               
    }                                                                                                     
}
```

**Comment by Notliam99** [comment]
I know this doesn't help anyone who isn't familiar with rust but there is a [niri_ipc](https://docs.rs/niri-ipc/latest/niri_ipc/) crate which could be used to watch for this event. This would work for any other complex event comparisons that are not possible currently.

**Comment by benjajaja** [comment]
@Notliam99 while that could be used to detect "single window on a workspace event", I don't see any way to toggle border/ring-focus with niri_ipc (nor `niri msg action` which seems to be the source of possible actions for niri_ipc).

**Reply by Notliam99** [reply]
There is a [way](https://github.com/YaLTeR/niri/discussions/773) but I gotta be honest it's pretty silly IPC have access to the parsed config.kdl if it did allow for plugins and the like.

**Comment by ranidspace** [comment]
I don't have much input here but, but no gaps or window borders on a centered, maximized window, would probably make niri a perfect compositor for me.

I would like no outer gaps with only one window, gaps seem to be global, the way hyprland does it sets the gaps for the workspace to be 0 if there's only one window on the workspace.

Perhaps maybe being able to set layout options for named workspaces could be good? If that's easy I think a workaround for a bit would be having a dedicated "maximized" workspace by turning off gaps and borders in it and putting whichever applications you need there? Kinda bad but _could_ be easier to implement

**Reply by YaLTeR** [reply]
Layout rules for named workspaces may happen yes

**Comment by ghost** [comment]
Is smart borders/gaps implemented, or is there a workaround? I use gaps and borders on my primary display but only ever have one window opened on my secondary displays and the gaps/borders are an eyesore. I used niri months ago but stopped soley because of this :(

**Comment by YaLTeR** [comment]
I'm implementing proper Wayland maximize here that removes gaps, borders and struts: https://github.com/YaLTeR/niri/pull/2376

**Reply by ranidspace** [reply]
This is the last thing I needed before I could switch to niri thank you for all the hard work being put into this, I cannot wait for it to get merged and released (actually i can, please take your time!)

**Comment by k4lizen** [comment]
My solution for this problem is that, after https://github.com/YaLTeR/niri/pull/2376 . I've rebound Mod+F to `maximize-window-to-edges`, and changed my `open-maximized` window rules to `open-maximized-to-edges`.

**Reply by k4lizen** [reply]
Although, my apps now have rounded corners when maximized like this. Im not sure why? Can I do something about this?

**Comment by 0pLuS0** [comment]
I was trying to remove the border around mpv, but not working...

window-rule {
    match app-id="mpv"
    open-floating true
     border {                                                                                              
        off 
  }
}

Do we have a rule for niri to remove the borders?

THANKS

**Reply by 0pLuS0** [reply]
After making this post, I got it working... :+1: 

window-rule {
    match app-id="mpv"
    open-floating true
    focus-ring {
    off
    }
}
